MSAL/src/instance/MSALAccountsProvider.m

```diff
@@ -178,21 +202,62 @@ - (MSALAccount *)accountForUsername:(NSString *)username
         return nil;
     }
     
+    return [self msalAccountsFromMSIDAccounts:msidAccounts];
+}
+
+- (NSArray<MSALAccount *> *)msalAccountsFromMSIDAccounts:(NSArray *)msidAccounts
+{
     NSMutableSet *msalAccounts = [NSMutableSet new];
     
     for (MSIDAccount *msidAccount in msidAccounts)
     {
-        MSALAccount *msalAccount = [[MSALAccount alloc] initWithMSIDAccount:msidAccount];
+        [self loadIdTokenClaimsForMSIDAccount:msidAccount];
+        
+        MSALAccount *msalAccount = [[MSALAccount alloc] initWithMSIDAccount:msidAccount createTenantProfile:YES];
+        if (!msalAccount) continue;
         
-        if (msalAccount)
+        MSALAccount *existAccount = [msalAccounts member:msalAccount];
+        if (!existAccount)
         {
             [msalAccounts addObject:msalAccount];
         }
+        else
+        {
+            [existAccount addTenantProfiles:msalAccount.tenantProfiles];
+        }
     }
     
     return [msalAccounts allObjects];
 }
 
+- (void)loadIdTokenClaimsForMSIDAccount:(MSIDAccount *)msidAccount
+{
+    // Return if id token claims is already loaded. This normally happen msidAccount retrieved from legacy tokens
+    if (msidAccount.idTokenClaims) return;
+    
+    MSIDConfiguration *config = [MSIDConfiguration new];
+    config.authority = msidAccount.authority;
+    config.clientId = self.clientId;
+    NSError *error;
+    MSIDIdToken *idToken = [self.tokenCache getIDTokenForAccount:msidAccount.accountIdentifier
```

Comment:
Are we going to do a separate keychain query for each account? This seems very costly. When we're reading accounts from default cache, we're already also reading refresh tokens. Can we also read id tokens similarly and merge based on some parameter for the internal API?

Reply:
addressed by doing id token query only once rather than every time.